lib/Dialect/LWE/Conversions/LWEToOpenfhe/LWEToOpenfhe.cpp

```diff
@@ -22,125 +30,286 @@
-                               "bigger than 64 bits.";
-
-    if (floatTy.getWidth() < 64) {
-      // OpenFHE has a convention that all inputs to MakeCKKSPackedPlaintext are
-      // std::vector<double>, so we need to cast the input to that type.
-      auto f64Ty = rewriter.getF64Type();
-      auto newTensorTy = RankedTensorType::get(tensorTy.getShape(), f64Ty);
-      input = rewriter.create<arith::ExtFOp>(op.getLoc(), newTensorTy, input);
+
+    // Cast inputs to the correct types for OpenFHE API.
+    if (auto intTy = mlir::dyn_cast<IntegerType>(elementTy)) {
+      if (intTy.getWidth() > 64)
+        return op.emitError() << "No supported packing technique for integers "
+                                 "bigger than 64 bits.";
+
+      if (intTy.getWidth() < 64) {
+        // OpenFHE has a convention that all inputs to MakePackedPlaintext are
+        // std::vector<int64_t>, so we need to cast the input to that type.
+        auto int64Ty = rewriter.getIntegerType(64);
+        auto newTensorTy = RankedTensorType::get(tensorTy.getShape(), int64Ty);
+        input =
+            rewriter.create<arith::ExtSIOp>(op.getLoc(), newTensorTy, input);
+      }
+    } else {
+      auto floatTy = cast<FloatType>(elementTy);
+      if (floatTy.getWidth() > 64)
+        return op.emitError() << "No supported packing technique for floats "
+                                 "bigger than 64 bits.";
+
+      if (floatTy.getWidth() < 64) {
+        // OpenFHE has a convention that all inputs to MakeCKKSPackedPlaintext
+        // are std::vector<double>, so we need to cast the input to that type.
+        auto f64Ty = rewriter.getF64Type();
+        auto newTensorTy = RankedTensorType::get(tensorTy.getShape(), f64Ty);
+        input = rewriter.create<arith::ExtFOp>(op.getLoc(), newTensorTy, input);
+      }
     }
-  }
 
-  lwe::NewLWEPlaintextType plaintextType = lwe::NewLWEPlaintextType::get(
-      op.getContext(),
-      lwe::ApplicationDataAttr::get(adaptor.getInput().getType(),
-                                    lwe::NoOverflowAttr::get(getContext())),
-      lwe::PlaintextSpaceAttr::get(getContext(), op.getRing(),
-                                   op.getEncoding()));
-  if (this->ckks_) {
-    rewriter.replaceOpWithNewOp<openfhe::MakeCKKSPackedPlaintextOp>(
-        op, plaintextType, cryptoContext, input);
-  } else {
-    rewriter.replaceOpWithNewOp<openfhe::MakePackedPlaintextOp>(
-        op, plaintextType, cryptoContext, input);
+    lwe::NewLWEPlaintextType plaintextType = lwe::NewLWEPlaintextType::get(
+        op.getContext(),
+        lwe::ApplicationDataAttr::get(adaptor.getInput().getType(),
+                                      lwe::NoOverflowAttr::get(getContext())),
+        lwe::PlaintextSpaceAttr::get(getContext(), op.getRing(),
+                                     op.getEncoding()));
+    if (isa<lwe::InverseCanonicalEncodingAttr>(op.getEncoding()) ||
```

Comment:
nit: this could be simplified with a type switch.

Reply:
Happy to change it if you want, but I don't think there's a nice way to handle the OR in that case? (I.e.., I don't think llvm::TypeSwitch supports fall-through?)

Reply:
TypeSwitch supports variadic template arguments in the match which function as a logical OR, see for example https://github.com/llvm/llvm-project/blob/a957cedea9657addbe8b860852cc98306aa437e7/mlir/lib/Dialect/LLVMIR/IR/LLVMTypes.cpp#L894-L898

The caveat is that you have to use `auto` for the callback's argument type and the template expansion has to compile. (or use a more generic type that encapsulates all options). I don't think that should be a problem for this code.

Reply:
Ohh, cool! Thanks for showing me the example "in the wild" - I actually briefly wondered about whether it does (because I knew isa<> does variadiac) but couldn't find any info on it in the docs (probably should have just tried but it was late...)

I'll add this and then squash/rebase :)